### PR TITLE
Feature: pdfplumber PDF loader with BaseBlobParser, without visual debugger

### PR DESCRIPTION
**Pull request by lesterpjy**
# Same as #4353 but with the reusable Parsers, and without the loader for visual debugger

#4353 but re-implemented with the new BaseBlobParser.
Method for saving a new file with visual debugger is removed as it does not fit well with rest of the loaders. Though by providing the option for using pdfplumber, users can check with the visual debugger themselves. 

## Who can review?

@eyurtsev 
@dev2049

### CODE REVIEW

File: langchain/document_loaders/pdf.py [hunk 1]
```diff
@@ -362,3 +363,26 @@ def load(self) -> List[Document]:
...
+            import pdfplumber  # noqa:F401
+        except ImportError:
+            raise ValueError(
+                "pdfplumber package not found, please install it with "
+                "`pip install pdfplumber`"
+            )
+
+        super().__init__(file_path)
+
+    def load(self, **kwargs: Optional[Any]) -> List[Document]:
```
Comment (eyurtsev) [inline langchain/document_loaders/pdf.py:383]: `load` should not take arguments or kwargs, so the interface remains `load()`

```suggestion
    def load(self) -> List[Document]:
```

All configuration differences get restricted to the __init__,  making it possible to substitute one loader for another in code that relies on loaders having a uniform load interface.

File: langchain/document_loaders/pdf.py [hunk 2]
```diff
@@ -362,3 +363,26 @@ def load(self) -> List[Document]:
             contents = self.clean_pdf(contents)
         metadata = {"source": self.source, "file_path": self.source}
         return [Document(page_content=contents, metadata=metadata)]
+
+
+class PDFPlumberLoader(BasePDFLoader):
+    """Loader that uses pdfplumber to load PDF files."""
+
+    def __init__(self, file_path: str) -> None:
```
Comment (eyurtsev) [inline langchain/document_loaders/pdf.py:371]:
```suggestion
    def __init__(self, file_path: str, text_kwargs: Optional[Mapping[str, Any]]=None) -> None:
```

File: langchain/document_loaders/pdf.py [hunk 3]
```diff
@@ -362,3 +363,26 @@ def load(self) -> List[Document]:
...
+        """Initialize with file path."""
+        try:
+            import pdfplumber  # noqa:F401
+        except ImportError:
+            raise ValueError(
+                "pdfplumber package not found, please install it with "
+                "`pip install pdfplumber`"
+            )
+
+        super().__init__(file_path)
```
Comment (eyurtsev) [inline langchain/document_loaders/pdf.py:381]:
```suggestion
        super().__init__(file_path)
        self.text_kwargs = text_kwargs or {}
```

File: langchain/document_loaders/pdf.py [hunk 4]
```diff
@@ -362,3 +363,26 @@ def load(self) -> List[Document]:
...
+
+    def load(self, **kwargs: Optional[Any]) -> List[Document]:
+        """Load file."""
+
+        parser = PDFPlumberParser(text_kwargs=kwargs)
```
Comment (eyurtsev) [inline langchain/document_loaders/pdf.py:386]:
```suggestion
        parser = PDFPlumberParser(text_kwargs=self.text_kwargs)
```

File: langchain/document_loaders/pdf.py (hunk 3 above)
Comment (eyurtsev) [inline langchain/document_loaders/pdf.py:381]:
```suggestion
        super().__init__(file_path)
        self.text_kwargs = text_kwargs or {}
```

File: tests/unit_tests/document_loader/parsers/test_pdf_parsers.py [hunk 5]
```diff
@@ -77,3 +78,9 @@ def test_pypdfium2_parser() -> None:
     """Test PyPDFium2 parser."""
     # Does not follow defaults to split by page.
     _assert_with_parser(PyPDFium2Parser())
+
+
+@pytest.mark.requires("pdfplumber")
```
Comment (eyurtsev) [inline tests/unit_tests/document_loader/parsers/test_pdf_parsers.py:83]: :heart:

**Comment by eyurtsev** [comment]
@lesterpjy  Thanks looking great! a few minor comments and then we can merge! Thanks for putting the time to work on this!

**Comment by lesterpjy** [comment]
Thanks for all the guidance! It's been a good learning experience for me. Looking forward to making more pull requests here.

**Comment by eyurtsev** [comment]
@lesterpjy I'm updating some of the files that you're also working on, which will be causing annoying merge conflicts for you. If you I can commandeer your PR from here to resolve the merge conflicts, and you can continue on more feature work. Let me know!

**Comment by lesterpjy** [comment]
@eyurtsev Sure, I'll try my best to resolve these conflicts. Will let you know if I have any questions or am facing any issues.

**Comment by eyurtsev** [comment]
@lesterpjy Thank you! :)

**Pull request merged**